wasm/validation.cpp

```diff
@@ -0,0 +1,520 @@
+        return tl::unexpected{ValidationError::ValueStackUnderflow};
+    }
+
+    auto val = value_stack.back();
+
+    value_stack.pop_back();
+
+    return val;
+}
+
+tl::expected<ValueOrUnknown, ValidationError> Validator::pop_val_expect(ValueOrUnknown const &expected) {
+    auto actual = pop_val();
+
+    if (!actual.has_value()) {
+        return tl::unexpected{actual.error()};
+    }
+
+    if (actual != expected && !std::holds_alternative<Unknown>(*actual) && !std::holds_alternative<Unknown>(expected)) {
+        return tl::unexpected{ValidationError::ValueStackUnexpected};
+    }
+
+    return actual;
+}
+
+void Validator::push_vals(std::vector<ValueType> const &vals) {
+    value_stack.reserve(value_stack.size() + vals.size());
+
+    for (ValueType const &val : vals) {
+        value_stack.push_back(val);
+    }
+}
+
+tl::expected<std::vector<ValueOrUnknown>, ValidationError> Validator::pop_vals(std::vector<ValueType> const &vals) {
+    std::vector<ValueOrUnknown> popped;
+
+    // TODO(dzero): switch to std::ranges::reverse_view once we drop older toolchains
+    // NOLINTNEXTLINE(modernize-loop-convert)
+    for (auto it = vals.rbegin(); it != vals.rend(); it++) {
+        auto maybe_val = pop_val_expect(*it);
+
+        if (!maybe_val.has_value()) {
+            return tl::unexpected{maybe_val.error()};
+        }
+
+        popped.insert(popped.begin(), *maybe_val);
+    }
+
+    return popped;
+}
+
+void Validator::push_ctrl(
+        Instruction const &i, std::vector<ValueType> const &params, std::vector<ValueType> const &results) {
+    control_stack.push_back(ControlFrame{i, params, results, value_stack.size(), false});
+
+    if (!params.empty()) {
+        push_vals(params);
+    }
+}
+
+tl::expected<ControlFrame, ValidationError> Validator::pop_ctrl() {
+    if (control_stack.empty()) {
+        return tl::unexpected{ValidationError::ControlStackEmpty};
+    }
+
+    auto frame = control_stack.back();
+
+    if (!frame.results.empty()) {
+        auto maybe_vals = pop_vals(frame.results);
+
+        if (!maybe_vals.has_value()) {
+            return tl::unexpected{maybe_vals.error()};
+        }
+    }
+
+    if (value_stack.size() != frame.stack_height) {
+        return tl::unexpected{ValidationError::ValueStackHeightMismatch};
+    }
+
+    control_stack.pop_back();
+
+    return frame;
+}
+
+std::vector<ValueType> const &Validator::label_types(ControlFrame &frame) {
+    if (std::holds_alternative<Loop>(frame.i)) {
+        return frame.params;
+    }
+
+    return frame.results;
+}
+
+void Validator::unreachable() {
+    value_stack.resize(control_stack.back().stack_height);
+
+    control_stack.back().unreachable = true;
+}
+
+// TODO(dzero): Serialize operand stack and control stack as part of the ValidationError to make debugging easier
+// https://webassembly.github.io/spec/core/valid/instructions.html#instruction-sequences
+tl::expected<bool, ValidationError> is_valid(CodeEntry const &ce, Module const &m) {
```

Comment:
Looks like this can only ever return `true` as the bool? Maybe `expected<void, ValidationError>` is a more suitable return type? I think that's pretty common for "returns nothing or fails"-functions.

Reply:
Didn't realize this was even possible.